### DISCUSSION TITLE
Querying foreign tables in a realtime subscription?

### DISCUSSION THREAD

**Discussion by madeleineostoja** [supabase/supabase#6059] [Questions] [answered]
I have a realtime subscription set up to sync local store in my app with some Supabase tables, and I can't find a way to fetch data from foreign key tables in that subscription? 

Eg: I have a table `photos` that has a foreign key `image` which links to the table `images`. My initial query to populate my store looks something like `db.from('photos').select('...', images(...)').eq('user_id:...')`, and that works great. But updates in a subscriber to `photos` can't get data from `images`. Am I missing something?

**Comment by w3b6x9** [comment, marked as answer]
@madeleineostoja A single Realtime subscription currently does not support listening to a table **and** its foreign key table for changes. A workaround for now, albeit not ideal, is to have two subscriptions each listening to your two tables and consolidating client-side.

**Reply by MaxwellGover** [reply]
@w3b6x9 How naughty would it be to just make a call to the db every time an insert is made? For example

```
function subscribeToComments() {
      supabase
      .channel('public:comments')
      .on('postgres_changes', { event: '*', schema: 'public', table: 'comments', filter: `chat_id=eq.${chat.id}` }, async (payload) => {
        let { data: newComment, error } = await supabase
          .from('comments')
          .select(`
            *,
            profiles (display_name, username), likes(*))
          `)
          .eq("id", payload.new.id)
          .maybeSingle();
  
        if (error) {
          console.log('ERROR', error.message)
        }
        setComments([...comments].concat([newComment]))
      })
      .subscribe()
    }
```

Here I need the profile and like information from another table but no changes are actually being made to those tables when a comment is added

**Reply by w3b6x9** [reply]
@MaxwellGover nice workaround and definitely something you can do as long as your database can support the frequency and query.

**Reply by andrevenancio** [reply]
Is this still the case in 2023?

**Comment by VittorioParagallo** [comment]
are realtime complex queries going to be supported?

**Reply by w3b6x9** [reply]
@VittorioParagallo it is not currently on the roadmap.

**Comment by Rubenburdin** [comment]
Still the case in 2025 unfortunately. Realtime subscriptions only return the changed row from that specific table, no foreign key joins.
@MaxwellGover's workaround is the standard pattern and works fine for most cases. You get the change notification, then immediately fetch the full record with relations:
```
`.on('postgres_changes', { event: 'INSERT', schema: 'public', table: 'photos' }, async (payload) => {
  const { data } = await supabase
    .from('photos')
    .select('*, images(*)')
    .eq('id', payload.new.id)
    .single();
  
  // use data with full relations
})`
```
The extra round trip adds maybe 20 to 50ms latency which is negligible for most UIs.
Alternative if you control the schema: denormalize the fields you need into the main table. Either duplicate them on insert, or use a generated column:
```
`alter table photos 
add column image_url text 
generated always as (
  (select url from images where images.id = photos.image)
) stored;`
```
Then Realtime returns that column directly. Tradeoff is storage and keeping it updated, but it avoids the second query.
For high frequency updates: multiple subscriptions (one per table) consolidated client side can actually perform better than firing a query on every change, since you're just receiving pushes rather than making requests.
No sign of joined realtime queries on the roadmap still. The WAL based CDC approach makes it tricky since Postgres streams row level changes, not query results.
